src/main/scala/org/camunda/feel/impl/interpreter/EvalContext.scala

```diff
@@ -35,77 +38,133 @@ class EvalContext(val valueMapper: ValueMapper,
   )
 
 }
 
 object EvalContext {
 
+  def empty(valueMapper: ValueMapper): EvalContext = create(
+    valueMapper = valueMapper,
+    functionProvider = EmptyFunctionProvider
+  )
+
+  def create(valueMapper: ValueMapper, functionProvider: FunctionProvider): EvalContext = new EvalContext(
+    valueMapper = valueMapper,
+    variableProvider = EmptyVariableProvider,
+    functionProvider = functionProvider
+  )
+
   def wrap(context: Context)(implicit valueMapper: ValueMapper): EvalContext =
-    new EvalContext(
-      valueMapper = valueMapper,
-      variableProvider = context.variableProvider,
-      functionProvider = context.functionProvider
-    )
+    context match {
+      case evalContext: EvalContext => evalContext
+      case EmptyContext => empty(valueMapper)
+      case StaticContext(variables, functions) => new EvalContext(
+        valueMapper = valueMapper,
+        variableProvider = toSortedVariableProvider(variables),
+        functionProvider = StaticFunctionProvider(functions)
+      )
+      case _ => new EvalContext(
+        valueMapper = valueMapper,
+        variableProvider = context.variableProvider,
+        functionProvider = context.functionProvider
+      )
+    }
+
+  private def mergeVariableProvider(provider: VariableProvider, otherProvider: VariableProvider): VariableProvider = {
+    (provider, otherProvider) match {
+      case (EmptyVariableProvider, EmptyVariableProvider) => EmptyVariableProvider
+      case (EmptyVariableProvider, otherProvider) => otherProvider
+      case (thisProvider, EmptyVariableProvider) => thisProvider
+      case (StaticVariableProvider(thisVariables), StaticVariableProvider(otherVariables)) =>
+        StaticVariableProvider(thisVariables ++ otherVariables)
+      case (thisProvider, otherProvider) => CompositeVariableProvider(List(thisProvider, otherProvider))
+    }
+  }
+
+  private def mergeFunctionProviders(provider: FunctionProvider, otherProvider: FunctionProvider): FunctionProvider = {
+    (provider, otherProvider) match {
+      case (EmptyFunctionProvider, EmptyFunctionProvider) => EmptyFunctionProvider
+      case (EmptyFunctionProvider, otherProvider) => otherProvider
+      case (thisProvider, EmptyFunctionProvider) => thisProvider
+      case (StaticFunctionProvider(thisFunctions), StaticFunctionProvider(otherFunctions)) => {
+        val allKeys = thisFunctions.keys ++ otherFunctions.keys
+        val functionsByKey = (key: String) => thisFunctions.getOrElse(key, List.empty) ++ otherFunctions.getOrElse(key, List.empty)
+        val allFunctions = allKeys.map(key => key -> functionsByKey(key)).toMap
+        StaticFunctionProvider(allFunctions)
+      }
+      case (thisProvider, otherProvider) => CompositeFunctionProvider(List(thisProvider, otherProvider))
+    }
+  }
+
+  private def toSortedVariableProvider(variables: Map[String, Any]): VariableProvider =
+    StaticVariableProvider(SortedMap[String, Any]() ++ variables)
```

Comment:
❌ As discussed f2f, this [SortedMap](https://www.scala-lang.org/api/2.13.6/scala/collection/immutable/SortedMap.html) does not behave the way you think it does. It doesn't keep the entries sorted in the order they were added. Instead they are sorted by their key, regardless of when they're added to the map.

Let's also add a test case for this scenario 🙂